File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.js
// your code goes here
const fs = require('fs');

let input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

input[0] = input[0].trim().split(' ');

let N = parseInt(input[0][0]);
let M = parseInt(input[0][1]);
let R = parseInt(input[0][2]);

let edges = [];

for (let i = 0; i <= N; i++) {
    edges.push([]);
}

for (let i = 1; i <= M; i++) {
    input[i] = input[i].trim().split(' ');
    let u = parseInt(input[i][0]);
    let v = parseInt(input[i][1]);
    
    edges[u].push(v);
    edges[v].push(u);
}

let queue = [];
let visited = Array.from({length: N + 1}, () => false);
let sequences = Array.from({length: N + 1}, () => 0);
let count = 1;

queue.push(R);
visited[R] = true;
sequences[R] = count++;

while (queue.length != 0) {
    let v = queue.shift();
    
    edges[v].sort((a, b) => a - b);
    
    for (let nv of edges[v]) {
        if (!visited[nv]) {
            queue.push(nv);
            visited[nv] = true;
            sequences[nv] = count++;
        }
    }
}

for (let i = 1; i <= N; i++) {
    console.log(sequences[i]);
}

File: 백준/Silver/24444. 알고리즘 수업 － 너비 우선 탐색 1/알고리즘 수업 － 너비 우선 탐색 1.cc
#include <iostream>
#include <queue>
#include <algorithm>

using namespace std;

int main() {
    int N, M, R;
    
    cin >> N >> M >> R;
    
    vector<int> edges[N + 1];
    
    for (int i = 0; i < M; i++) {
        int u, v;
        
        cin >> u >> v;
        
        edges[u].push_back(v);
        edges[v].push_back(u);
    }
    
    queue<int> queue;
    bool visited[N + 1]{false};
    int sequence[N + 1]{0};
    int count = 1;
    
    queue.push(R);
    visited[R] = true;
    sequence[R] = count++;
    
    while (!queue.empty()) {
        int vertex = queue.front();
        queue.pop();
        
        sort(edges[vertex].begin(), edges[vertex].end(), less<int>());
        for (int nextVertex : edges[vertex]) {
            if (!visited[nextVertex]) {
                queue.push(nextVertex);
                visited[nextVertex] = true;
                sequence[nextVertex] = count++;
            }
        }
    }
    
    for (int i = 1; i <= N; i++) {
        cout << sequence[i] << '\n';
    }
}